### PR TITLE
Summarize model as rich table

### PR DESCRIPTION
**Pull request by ricardoV94**
From https://github.com/pymc-devs/pymc/discussions/7210#discussioncomment-10921154

```python
import numpy as np
import pymc as pm

from pymc_experimental.printing import model_table

coords = {"subject": range(20), "param": ["a", "b"]}
with pm.Model(coords=coords) as m:
    x = pm.Data("x", np.random.normal(size=(20, 2)), dims=("subject", "param"))
    y = pm.Data("y", np.random.normal(size=(20,)), dims="subject")

    beta = pm.Normal("beta", mu=0, sigma=1, dims="param")
    mu = pm.Deterministic("mu", pm.math.dot(x, beta), dims="subject")
    sigma = pm.HalfNormal("sigma", sigma=1)

    y_obs = pm.Normal("y_obs", mu=mu, sigma=sigma, observed=y, dims="subject")

model_table(m)
```
```
 Variable  Expression         Dimensions             
─────────────────────────────────────────────────────
      x =  Data               subject[20] × param[2] 
      y =  Data               subject[20]            
                                                     
   beta ~  Normal(0, 1)       param[2]               
  sigma ~  HalfNormal(0, 1)                          
                              Parameter count = 3    
                                                     
     mu =  f(beta)            subject[20]            
                                                     
  y_obs ~  Normal(mu, sigma)  subject[20]     
```

### CODE REVIEW

**Comment by ricardoV94** [comment]
The output will be automatically improved once these issues are solved:
1. https://github.com/pymc-devs/pymc/issues/7536
1. https://github.com/pymc-devs/pymc/issues/7538

File: pymc_experimental/printing.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+
+def _extract_value(var: SharedVariable | Constant) -> np.ndarray:
+    if isinstance(var, SharedVariable):
+        return var.get_value(borrow=True)
+    else:
+        return var.data
+
+
+def model_table(
```
Comment (jessegrabowski) [inline pymc_experimental/printing.py:20]: I would suggest splitting out the functionality for actually creating the string representation from this function. That intermediate representation could return a list of (name, expression, dims) tuples, which could then be fed into different types of string representation (of which `model_table` is one)

Comment (ricardoV94) [reply pymc_experimental/printing.py:20]: Yeah I was thinking the same

Comment (ricardoV94) [reply pymc_experimental/printing.py:20]: I refactored the utilities that convert model info (variable expressions, dims) so they can be used elsewhere

**Review by jessegrabowski — APPROVED** [review]

**Pull request merged**